Add a support for immovable generators

For example, when I try to format the following code,

```rust
#![feature(generators)]

unsafe fn foo() {
    let mut g = static || {
        yield 3;
    };
}
```

rustfmt 0.3.6-nightly emits the following error:
```
error: expected expression, found keyword `static`
 --> test.rs:4:17
  |
4 |     let mut g = static || {
  |                 ^^^^^^

```

The procedure to fix this would be something like this:
1. add tests to make sure that rustfmt handles immovable generators
2. update `rustc-ap-syntax` and `rustc-ap-rustc_errors` to the latest version
3. update expr.rs and closures.rs to support immovable generators

cc https://github.com/rust-lang/rust/pull/45337

I have updated `rustc-ap-syntax` and `rustc-ap-rustc_errors` to the latest version.
But the latest `rustc-ap-syntax = "26.0.0"` introduced a new struct inside `ast.rs`:
```
pub struct Label {
    pub ident: Ident,
    pub span: Span,
}
```
to replace the function provided by `syntax::codemap::Spanned` in `rustc-ap-syntax = "12.0.0"`.
So it broke some piece of code: 
```
error[E0609]: no field `node` on type `syntax::ast::Label`
   --> src/expr.rs:140:53
    |
140 |                 Some(ident) => format!(" {}", ident.node),
    |                                                     ^^^^ unknown field
    |
    = note: available fields are: `ident`, `span`

error[E0609]: no field `node` on type `syntax::ast::Label`
   --> src/expr.rs:147:53
    |
147 |                 Some(ident) => format!(" {}", ident.node),
    |                                                     ^^^^ unknown field
    |
    = note: available fields are: `ident`, `span`

error[E0277]: the trait bound `str: std::marker::Sized` is not satisfied
   --> src/expr.rs:297:20
    |
297 |         .and_then(|expr_str| {
    |                    ^^^^^^^^ `str` does not have a constant size known at compile-time
    |
    = help: the trait `std::marker::Sized` is not implemented for `str`
    = note: all local variables must have a statically known size

error[E0308]: mismatched types
   --> src/expr.rs:755:57
    |
755 |             Some(ControlFlow::new_for(pat, cond, block, label, expr.span))
    |                                                         ^^^^^ expected struct `syntax::codemap::Spanned`, found struct `syntax::ast::Label`
    |
    = note: expected type `std::option::Option<syntax::codemap::Spanned<syntax::ast::Ident>>`
               found type `std::option::Option<syntax::ast::Label>`
```
It can be fixed with some modifications, is it ok to enclose these modifications into commits for this issue(I'm uncertain because this bug is introduced by `rust-ap-syntax` but unrelated to `immovable_generator`)?

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -44,8 +44,8 @@ env_logger = "0.4"
 getopts = "0.2"
 derive-new = "0.5"
 cargo_metadata = "0.4"
-rustc-ap-syntax = "12.0.0"
-rustc-ap-rustc_errors = "12.0.0"
+rustc-ap-syntax = "26.0.0"
+rustc-ap-rustc_errors = "26.0.0"
 
 [dev-dependencies]
 lazy_static = "1.0.0"

--- a/src/closures.rs
+++ b/src/closures.rs
@@ -278,7 +278,7 @@ pub fn rewrite_last_closure(
     expr: &ast::Expr,
     shape: Shape,
 ) -> Option<String> {
-    if let ast::ExprKind::Closure(capture, ref fn_decl, ref body, _) = expr.node {
+    if let ast::ExprKind::Closure(capture, _, ref fn_decl, ref body, _) = expr.node {
         let body = match body.node {
             ast::ExprKind::Block(ref block)
                 if !is_unsafe_block(block) && is_simple_block(block, context.codemap) =>

--- a/src/closures.rs
+++ b/src/closures.rs
@@ -33,6 +33,7 @@ use utils::{last_line_width, left_most_sub_expr, stmt_expr};
 
 pub fn rewrite_closure(
     capture: ast::CaptureBy,
+    movability: ast::Movability,
     fn_decl: &ast::FnDecl,
     body: &ast::Expr,
     span: Span,
@@ -42,7 +43,7 @@ pub fn rewrite_closure(
     debug!("rewrite_closure {:?}", body);
 
     let (prefix, extra_offset) =
-        rewrite_closure_fn_decl(capture, fn_decl, body, span, context, shape)?;
+        rewrite_closure_fn_decl(capture, movability, fn_decl, body, span, context, shape)?;
     // 1 = space between `|...|` and body.
     let body_shape = shape.offset_left(extra_offset)?;
 
@@ -194,6 +195,7 @@ fn rewrite_closure_block(
 // Return type is (prefix, extra_offset)
 fn rewrite_closure_fn_decl(
     capture: ast::CaptureBy,
+    movability: ast::Movability,
     fn_decl: &ast::FnDecl,
     body: &ast::Expr,
     span: Span,
@@ -205,9 +207,17 @@ fn rewrite_closure_fn_decl(
     } else {
         ""
     };
+
+    let immovable = if movability == ast::Movability::Static {
+        "static "
+    } else {
+        ""
+    };
     // 4 = "|| {".len(), which is overconservative when the closure consists of
     // a single expression.
-    let nested_shape = shape.shrink_left(mover.len())?.sub_width(4)?;
+    let nested_shape = shape
+        .shrink_left(mover.len() + immovable.len())?
+        .sub_width(4)?;
 
     // 1 = |
     let argument_offset = nested_shape.indent + 1;
@@ -254,7 +264,7 @@ fn rewrite_closure_fn_decl(
         config: context.config,
     };
     let list_str = write_list(&item_vec, &fmt)?;
-    let mut prefix = format!("{}|{}|", mover, list_str);
+    let mut prefix = format!("{}{}|{}|", immovable, mover, list_str);
 
     if !ret_str.is_empty() {
         if prefix.contains('\n') {
@@ -278,7 +288,7 @@ pub fn rewrite_last_closure(
     expr: &ast::Expr,
     shape: Shape,
 ) -> Option<String> {
-    if let ast::ExprKind::Closure(capture, _, ref fn_decl, ref body, _) = expr.node {
+    if let ast::ExprKind::Closure(capture, movability, ref fn_decl, ref body, _) = expr.node {
         let body = match body.node {
             ast::ExprKind::Block(ref block)
                 if !is_unsafe_block(block) && is_simple_block(block, context.codemap) =>
@@ -287,8 +297,15 @@ pub fn rewrite_last_closure(
             }
             _ => body,
         };
-        let (prefix, extra_offset) =
-            rewrite_closure_fn_decl(capture, fn_decl, body, expr.span, context, shape)?;
+        let (prefix, extra_offset) = rewrite_closure_fn_decl(
+            capture,
+            movability,
+            fn_decl,
+            body,
+            expr.span,
+            context,
+            shape,
+        )?;
         // If the closure goes multi line before its body, do not overflow the closure.
         if prefix.contains('\n') {
             return None;

--- a/src/expr.rs
+++ b/src/expr.rs
@@ -159,8 +159,16 @@ pub fn format_expr(
         } else {
             Some("yield".to_string())
         },
-        ast::ExprKind::Closure(capture, _, ref fn_decl, ref body, _) => {
-            closures::rewrite_closure(capture, fn_decl, body, expr.span, context, shape)
+        ast::ExprKind::Closure(capture, movability, ref fn_decl, ref body, _) => {
+            closures::rewrite_closure(
+                capture,
+                movability,
+                fn_decl,
+                body,
+                expr.span,
+                context,
+                shape,
+            )
         }
         ast::ExprKind::Try(..)
         | ast::ExprKind::Field(..)

--- a/src/expr.rs
+++ b/src/expr.rs
@@ -135,16 +135,16 @@ pub fn format_expr(
         ast::ExprKind::AssignOp(ref op, ref lhs, ref rhs) => {
             rewrite_assignment(context, lhs, rhs, Some(op), shape)
         }
-        ast::ExprKind::Continue(ref opt_ident) => {
-            let id_str = match *opt_ident {
-                Some(ident) => format!(" {}", ident.node),
+        ast::ExprKind::Continue(ref opt_label) => {
+            let id_str = match *opt_label {
+                Some(label) => format!(" {}", label.ident),
                 None => String::new(),
             };
             Some(format!("continue{}", id_str))
         }
-        ast::ExprKind::Break(ref opt_ident, ref opt_expr) => {
-            let id_str = match *opt_ident {
-                Some(ident) => format!(" {}", ident.node),
+        ast::ExprKind::Break(ref opt_label, ref opt_expr) => {
+            let id_str = match *opt_label {
+                Some(label) => format!(" {}", label.ident),
                 None => String::new(),
             };
 
@@ -159,7 +159,7 @@ pub fn format_expr(
         } else {
             Some("yield".to_string())
         },
-        ast::ExprKind::Closure(capture, ref fn_decl, ref body, _) => {
+        ast::ExprKind::Closure(capture, _, ref fn_decl, ref body, _) => {
             closures::rewrite_closure(capture, fn_decl, body, expr.span, context, shape)
         }
         ast::ExprKind::Try(..)
@@ -718,7 +718,7 @@ struct ControlFlow<'a> {
     cond: Option<&'a ast::Expr>,
     block: &'a ast::Block,
     else_block: Option<&'a ast::Expr>,
-    label: Option<ast::SpannedIdent>,
+    label: Option<ast::Label>,
     pat: Option<&'a ast::Pat>,
     keyword: &'a str,
     matcher: &'a str,
@@ -795,11 +795,7 @@ impl<'a> ControlFlow<'a> {
         }
     }
 
-    fn new_loop(
-        block: &'a ast::Block,
-        label: Option<ast::SpannedIdent>,
-        span: Span,
-    ) -> ControlFlow<'a> {
+    fn new_loop(block: &'a ast::Block, label: Option<ast::Label>, span: Span) -> ControlFlow<'a> {
         ControlFlow {
             cond: None,
             block: block,
@@ -819,7 +815,7 @@ impl<'a> ControlFlow<'a> {
         pat: Option<&'a ast::Pat>,
         cond: &'a ast::Expr,
         block: &'a ast::Block,
-        label: Option<ast::SpannedIdent>,
+        label: Option<ast::Label>,
         span: Span,
     ) -> ControlFlow<'a> {
         ControlFlow {
@@ -844,7 +840,7 @@ impl<'a> ControlFlow<'a> {
         pat: &'a ast::Pat,
         cond: &'a ast::Expr,
         block: &'a ast::Block,
-        label: Option<ast::SpannedIdent>,
+        label: Option<ast::Label>,
         span: Span,
     ) -> ControlFlow<'a> {
         ControlFlow {
@@ -1166,9 +1162,9 @@ impl<'a> Rewrite for ControlFlow<'a> {
     }
 }
 
-fn rewrite_label(label: Option<ast::SpannedIdent>) -> Cow<'static, str> {
-    match label {
-        Some(ident) => Cow::from(format!("{}: ", ident.node)),
+fn rewrite_label(opt_label: Option<ast::Label>) -> Cow<'static, str> {
+    match opt_label {
+        Some(label) => Cow::from(format!("{}: ", label.ident)),
         None => Cow::from(""),
     }
 }

--- a/tests/source/immovable_generators.rs
+++ b/tests/source/immovable_generators.rs
@@ -1,0 +1,7 @@
+#![feature(generators)]
+
+unsafe fn foo() {
+    let mut ga = static || { 
+        yield 1;
+    };
+}

--- a/tests/target/immovable_generators.rs
+++ b/tests/target/immovable_generators.rs
@@ -1,0 +1,7 @@
+#![feature(generators)]
+
+unsafe fn foo() {
+    let mut ga = static || {
+        yield 1;
+    };
+}

